execute_jscore.c

```diff
@@ -376,13 +379,31 @@ bool proxy_execute_jscore_global_init(void) {
     g_proxy_execute_jscore.module = dlopen(
         "/System/Library/Frameworks/JavaScriptCore.framework/Versions/Current/JavaScriptCore", RTLD_LAZY | RTLD_LOCAL);
 #else
-    g_proxy_execute_jscore.module = dlopen("libjavascriptcoregtk-4.1.so.0", RTLD_LAZY | RTLD_LOCAL);
-    if (!g_proxy_execute_jscore.module)
-        g_proxy_execute_jscore.module = dlopen("libjavascriptcoregtk-4.0.so.18", RTLD_LAZY | RTLD_LOCAL);
-    if (!g_proxy_execute_jscore.module)
-        g_proxy_execute_jscore.module = dlopen("libjavascriptcoregtk-3.0.so.0", RTLD_LAZY | RTLD_LOCAL);
-    if (!g_proxy_execute_jscore.module)
-        g_proxy_execute_jscore.module = dlopen("libjavascriptcoregtk-1.0.so.0", RTLD_LAZY | RTLD_LOCAL);
+    const char *library_names[] = {"libjavascriptcoregtk-4.1.so.0", "libjavascriptcoregtk-4.0.so.18",
+                                   "libjavascriptcoregtk-3.0.so.0", "libjavascriptcoregtk-1.0.so.0"};
+    const size_t library_names_size = sizeof(library_names) / sizeof(library_names[0]);
+
+    // Use existing JavaScriptCoreGTK if already loaded
+    struct link_map *map = NULL;
+    void *current_process = dlopen(0, RTLD_LAZY);
+    if (current_process) {
+        if (dlinfo(current_process, RTLD_DI_LINKMAP, &map) == 0) {
+            while (map && !g_proxy_execute_jscore.module) {
+                for (size_t i = 0; i < library_names_size; i++) {
+                    if (strstr(map->l_name, library_names[i])) {
+                        g_proxy_execute_jscore.module = dlopen(map->l_name, RTLD_NOLOAD | RTLD_LAZY | RTLD_LOCAL);
+                        break;
+                    }
+                }
+                map = map->l_next;
+            }
+        }
```

Comment:
Extremely deep indentations. Can you make a function that exits early if conditions are nor met.

Reply:
I think doing a separate function might make it more complex. I have removed one of the indentations though.